## Importing Libraries

In [1]:
import numpy as np 
import pandas as pd 
import csv 
pd.pandas.set_option('display.max_columns',None)
import warnings
warnings.filterwarnings('ignore')

import matplotlib.pyplot as plt
import seaborn as sns
import re
import string

import nltk
import random
from nltk.tokenize import RegexpTokenizer
from stop_words import get_stop_words
from nltk.stem.porter import PorterStemmer
from nltk.stem import WordNetLemmatizer 

from langdetect import detect

## Reading Data

In [2]:
data1=pd.read_csv("data/sf_aliens.csv")
data2=pd.read_csv("data/sf_alternate_history.csv")
data3=pd.read_csv("data/sf_alternate_universe.csv")
data4=pd.read_csv("data/sf_apocalyptic.csv")
data5=pd.read_csv("data/sf_cyberpunk.csv")
data6=pd.read_csv("data/sf_dystopia.csv")
data7=pd.read_csv("data/sf_hard.csv")
data8=pd.read_csv("data/sf_military.csv")
data9=pd.read_csv("data/sf_robots.csv")
data10=pd.read_csv("data/sf_space_opera.csv")
data11=pd.read_csv("data/sf_steampunk.csv")
data12=pd.read_csv("data/sf_time_travel.csv")

In [3]:
data1.head()

Book_Title Original_Book_Title             Author_Name Edition_Language  \
0      Obsidian            Obsidian  Jennifer L. Armentrout          English   
1          Onyx                Onyx  Jennifer L. Armentrout          English   
2  The 5th Wave        The 5th Wave             Rick Yancey          English   
3      The Host            The Host         Stephenie Meyer          English   
4          Opal                Opal  Jennifer L. Armentrout             None   

   Rating_score  Rating_votes  Review_number  \
0          4.17        236780          18161   
1          4.27        153429          10497   
2          4.03        400600          29990   
3          3.84        915026          41673   
4          4.27        129006           9463   

                                    Book_Description  Year_published  \
0  Starting over sucks.When we moved to West Virg...            2011   
1  BEING CONNECTED TO DAEMON BLACK SUCKS… Thanks ...            2012   
2  After the 1st wave, only darkness remains. Aft...            2013   
3  Melanie Stryder refuses to fade away. The eart...            2008   
4  No one is like Daemon Black.When he set out to...            2012   

                                              Genres  \
0  {'Young Adult': 3439, 'Fantasy (Paranormal) ':...   
1  {'Young Adult': 2271, 'Fantasy (Paranormal) ':...   
2  {'Young Adult': 5436, 'Science Fiction': 3327,...   
3  {'Young Adult': 4529, 'Science Fiction': 4285,...   
4  {'Young Adult': 1855, 'Fantasy (Paranormal) ':...   

                                                 url  
0  https://www.goodreads.com/book/show/12578077-o...  
1  https://www.goodreads.com/book/show/13047090-onyx  
2  https://www.goodreads.com/book/show/16101128-t...  
3  https://www.goodreads.com/book/show/1656001.Th...  
4  https://www.goodreads.com/book/show/13362536-opal

In [4]:
data2.tail()

Book_Title  \
1244  The Lady's Guide to Petticoats and Piracy   
1245                                        Vox   
1246               Arabella The Traitor of Mars   
1247                                      Circe   
1248                        An Ocean of Minutes   

                                    Original_Book_Title        Author_Name  \
1244          The Lady's Guide to Petticoats and Piracy       Mackenzi Lee   
1245                                                Vox  Christina Dalcher   
1246  \n                  0765382830\n              ...    David D. Levine   
1247                                              Circe    Madeline Miller   
1248                                An Ocean of Minutes           Thea Lim   

     Edition_Language  Rating_score  Rating_votes  Review_number  \
1244          English          4.09         25718           4777   
1245          English          3.55         55972           9070   
1246          English          3.77           205             44   
1247          English          4.26        392155          42746   
1248          English          3.45          4907            901   

                                       Book_Description  Year_published  \
1244  In this highly anticipated sequel to the New Y...            2018   
1245  Set in an America where half the population ha...            2018   
1246  Taking up almost immediately after the great B...            2018   
1247  In the house of Helios, god of the sun and mig...            2018   
1248  Shortlisted for the 2018 Scotiabank Giller Pri...            2018   

                                                 Genres  \
1244  {'Historical (Historical Fiction) ': 1559, 'Yo...   
1245  {'Fiction': 959, 'Science Fiction (Dystopia) '...   
1246  {'Science Fiction (Steampunk) ': 26, 'Science ...   
1247  {'Fantasy': 9508, 'Fiction': 5752, 'Fantasy (M...   
1248  {'Fiction': 193, 'Science Fiction (Time Travel...   

                                                    url  
1244  https://www.goodreads.com/book/show/37880094-t...  
1245   https://www.goodreads.com/book/show/37796866-vox  
1246  https://www.goodreads.com/book/show/36314198-a...  
1247  https://www.goodreads.com/book/show/35959740-c...  
1248  https://www.goodreads.com/book/show/42202089-a...

Concatenating all the 12 dataframes as they have same number or 11 columns

In [5]:
data=pd.concat([data1,data2,data3,data4,data5,data6,data7,data8,data9,\
          data10,data11,data12],axis=0)

In [6]:
print(f"The dataset has :\n{data.shape[0]} rows\n{data.shape[1]} columns")

The dataset has :
14974 rows
11 columns


The dataframe finally consists of 14974 rows.

Check for the null values in the dataframe containing all books

In [7]:
data.isnull().sum()

Book_Title              0
Original_Book_Title     0
Author_Name             0
Edition_Language       32
Rating_score            0
Rating_votes            0
Review_number           0
Book_Description       71
Year_published          0
Genres                  0
url                     0
dtype: int64

Since Edition consists of 32 Null values and Book_Description consits of 72 Null values out of 14974, we can drop these.

In [8]:
data.dropna(subset=['Edition_Language','Book_Description'],inplace=True)

In [9]:
data.isnull().sum()

Book_Title             0
Original_Book_Title    0
Author_Name            0
Edition_Language       0
Rating_score           0
Rating_votes           0
Review_number          0
Book_Description       0
Year_published         0
Genres                 0
url                    0
dtype: int64

In [10]:
print(f"The updated dataset has :\n{data.shape[0]} rows\n{data.shape[1]} columns")

The updated dataset has :
14871 rows
11 columns


Let's give ID to each of these books.\
Will generate numbers ranging form 0 to length of dataset and give it.
|

In [11]:
data['ID']=[i for i in range(0,data.shape[0])]

In [12]:
data.set_index('ID',inplace=True)

In [13]:
data.tail()

Book_Title    Original_Book_Title         Author_Name  \
ID                                                                        
14866  Tumbling Through Time  Tumbling Through Time         Gwyn Cready   
14867           Time's Arrow           Time’s Arrow         Martin Amis   
14868      Time Out of Joint      Time Out of Joint      Philip K. Dick   
14869              Fog Magic              Fog Magic      Julia L. Sauer   
14870     Farnham's Freehold     Farnham's Freehold  Robert A. Heinlein   

      Edition_Language  Rating_score  Rating_votes  Review_number  \
ID                                                                  
14866          English          3.34           312             51   
14867          English          3.71         14934           1106   
14868          English          3.85         11126            688   
14869          English          3.88          1204            170   
14870          English          3.56         11337            407   

                                        Book_Description  Year_published  \
ID                                                                         
14866  In her sparkling and hilarious debut, author G...            2000   
14867  In Time's Arrow the doctor Tod T. Friendly die...            1991   
14868  Time Out of Joint is Philip K. Dick’s classic ...            1959   
14869  Originally published in 1943, this edition fea...            1943   
14870  Hugh Farnham was a practical, self-made man, a...            1964   

                                                  Genres  \
ID                                                         
14866  {'Romance': 28, 'Science Fiction (Time Travel)...   
14867  {'Fiction': 586, 'Historical (Historical Ficti...   
14868  {'Science Fiction': 532, 'Fiction': 211, 'Nove...   
14869  {'Fantasy': 73, 'Fiction': 31, 'Childrens': 31...   
14870  {'Science Fiction': 579, 'Fiction': 131, 'Apoc...   

                                                     url  
ID                                                        
14866  https://www.goodreads.com/book/show/1888734.Tu...  
14867  https://www.goodreads.com/book/show/23031.Time...  
14868  https://www.goodreads.com/book/show/698034.Tim...  
14869  https://www.goodreads.com/book/show/297891.Fog...  
14870  https://www.goodreads.com/book/show/50840.Farn...

Let's see what exactly is in the genre column.\
These are the genres that the users marked for the book.\
So, it is dictionary with key as genre and value as the no. of users who makred the specific genre.\
An example of genres :


In [14]:
data.Genres[7]

"{'Young Adult': 3598, 'Fantasy': 2417, 'Science Fiction': 2033, 'Fiction': 936, 'Science Fiction (Aliens) ': 750, 'Romance': 479, 'Fantasy (Paranormal) ': 447, 'Adventure': 420, 'Young Adult (Teen) ': 263, 'Action': 241}"

In [15]:
data.Genres[2000]

"{'Sequential Art (Graphic Novels) ': 336, 'Sequential Art (Comics) ': 298, 'Fantasy': 87, 'Science Fiction': 74, 'Westerns': 51, 'Comics (Comic Book) ': 36, 'Fiction': 35, 'Graphic Novels Comics': 35, 'Science Fiction (Dystopia) ': 20, 'Horror': 17}"

In [16]:
data.Genres[3500]

"{'Fantasy': 3889, 'Young Adult': 1055, 'Fantasy (Magic) ': 639, 'Romance': 569, 'Fiction': 332, 'Adventure': 220, 'Fantasy (High Fantasy) ': 187, 'Fantasy (Paranormal) ': 161, 'Young Adult (Young Adult Fantasy) ': 142, 'Adult': 95}"

In [17]:
data.Genres[1400]

"{'Fantasy': 347, 'Fiction': 214, 'Science Fiction': 109, 'Historical (Historical Fiction) ': 77, 'Science Fiction (Alternate History) ': 72, 'Fantasy (Weird Fiction) ': 42, 'Weird Fiction (New Weird) ': 36, 'Speculative Fiction': 30, 'Historical': 29, 'Novella': 29}"

The dictionary is in descending order of values.\
So, we can consider taking 3-4 top genres form the dictionary.\
Let's see in a while how it works

### Are there duplicate values?

In [18]:
data.duplicated().sum()/data.shape[0]*100

15.923609710174164

There are 2368 duplicate values(around 16%)\
Let's drop the duplicates

In [19]:
data.drop_duplicates(inplace=True)

Let's see how many copies of the same book exist after dropping duplicates.\
Hunger games, for example :

In [20]:
data[data.Book_Title=='The Hunger Games']

Book_Title Original_Book_Title      Author_Name Edition_Language  \
ID                                                                             
1537  The Hunger Games    The Hunger Games  Suzanne Collins          English   
6206  The Hunger Games    The Hunger Games  Suzanne Collins          English   
8151  The Hunger Games    The Hunger Games  Suzanne Collins          English   

      Rating_score  Rating_votes  Review_number  \
ID                                                
1537          4.33       6572148         174079   
6206          4.33       6570886         174054   
8151          4.33       6572958         174090   

                                       Book_Description  Year_published  \
ID                                                                        
1537  Could you survive on your own in the wild, wit...            2008   
6206  Could you survive on your own in the wild, wit...            2008   
8151  Could you survive on your own in the wild, wit...            2008   

                                                 Genres  \
ID                                                        
1537  {'Young Adult': 31089, 'Fiction': 17571, 'Scie...   
6206  {'Young Adult': 31061, 'Fiction': 17561, 'Scie...   
8151  {'Young Adult': 31089, 'Fiction': 17571, 'Scie...   

                                                    url  
ID                                                       
1537  https://www.goodreads.com/book/show/2767052-th...  
6206  https://www.goodreads.com/book/show/2767052-th...  
8151  https://www.goodreads.com/book/show/2767052-th...

We have 3 "Hunger games" books. If we notice, there is a difference in the Rating_votes, Review_number, Genre. \
The Last one seems to be the most updated one.\
We'll check if we are able to notice the same pattern

In [21]:
data[data.Book_Title=='Dauntless']

Book_Title             Original_Book_Title    Author_Name  \
ID                                                                
7933   Dauntless  Dauntless (The Lost Fleet, #1)  Jack Campbell   
8664   Dauntless  Dauntless (The Lost Fleet, #1)  Jack Campbell   
11176  Dauntless  Dauntless (The Lost Fleet, #1)  Jack Campbell   

      Edition_Language  Rating_score  Rating_votes  Review_number  \
ID                                                                  
7933           English          3.96         27827           1585   
8664           English          3.96         27813           1584   
11176          English          3.96         27824           1584   

                                        Book_Description  Year_published  \
ID                                                                         
7933   The Alliance has been fighting the Syndics for...            2006   
8664   The Alliance has been fighting the Syndics for...            2006   
11176  The Alliance has been fighting the Syndics for...            2006   

                                                  Genres  \
ID                                                         
7933   {'Science Fiction': 1099, 'Space (Space Opera)...   
8664   {'Science Fiction': 1098, 'Space (Space Opera)...   
11176  {'Science Fiction': 1099, 'Space (Space Opera)...   

                                                     url  
ID                                                        
7933   https://www.goodreads.com/book/show/112292.Dau...  
8664   https://www.goodreads.com/book/show/112292.Dau...  
11176  https://www.goodreads.com/book/show/112292.Dau...

Doesn't necessarily mean last one has max value. We can drop the other and keep the one having the largest Rating_votes.\
But how many books will we end up with after that?

In [22]:
data.Book_Title.nunique()

10615

We should end up with **10615** books.\
Sort the Rating_votes in descending order and then remove te duplicate ones. 

In [23]:
data=data.sort_values('Rating_votes',ascending=False).\
drop_duplicates(subset=['Book_Title'])

data.shape

(10615, 11)

Works about right.

### How many languages?

In [24]:
data.Edition_Language.value_counts()

English     9855
None         653
German        33
French        21
Japanese      16
Polish        11
Russian        9
Spanish        6
Chinese        4
Dutch          2
Italian        2
Arabic         1
Swedish        1
Finnish        1
Name: Edition_Language, dtype: int64

In [25]:
data[~((data.Edition_Language=='English') | (data.Edition_Language=='None'))].shape

(107, 11)

We are focusing on English books right now, so we will drop anything that isn't English

In [26]:
data.drop(data.index[~((data.Edition_Language=='English') | (data.Edition_Language=='None'))],\
         axis=0,inplace=True)

In [27]:
data.shape

(10508, 11)

We now have English and None. Let's find out how many of None are English

We will be able to this only if we clean the text.

In [28]:
def clean_text(text):
    """
    function to clean the apostrophe's in the text
    for better language detection
    """
    text=re.sub(r'["]',r' ',text)
    text=re.sub("'"," ",text)
    return text

Creating a none datframe which have all languages as None

In [29]:
none=data[(data.Edition_Language=='None')][['Book_Title','Book_Description']]

In [30]:
none.head()

Book_Title                                   Book_Description
ID                                                                   
2717       Reached  After leaving Society to desperately seek The ...
4             Opal  No one is like Daemon Black.When he set out to...
4265  The Dreamers  In an isolated college town in the hills of So...
7104      The Gift  When Whit & Wisty were imprisoned by the wicke...
6716    Queen Song  Queen Coriane, first wife of King Tiberias, ke...

In [31]:
none.Book_Description=none.Book_Description.map(clean_text)

In [32]:
none.head()

Book_Title                                   Book_Description
ID                                                                   
2717       Reached  After leaving Society to desperately seek The ...
4             Opal  No one is like Daemon Black.When he set out to...
4265  The Dreamers  In an isolated college town in the hills of So...
7104      The Gift  When Whit & Wisty were imprisoned by the wicke...
6716    Queen Song  Queen Coriane, first wife of King Tiberias, ke...

In [33]:
none['lang']=none.Book_Description.map(detect)

In [34]:
none[none.lang!='en']

Book_Title  \
ID                                   
8268                Rogue in Space   
8018                      Poseidon   
8021  Phobos: Hard Science Fiction   
8455               Into the Future   

                                       Book_Description lang  
ID                                                            
8268  Ecco un originalissimo romanzo di Fredric Brow...   it  
8018  Der Wettlauf um die Eroberung des Sonnensystem...   de  
8021  Nach mehreren dramatischen Fehlschlägen nähert...   de  
8455                                       intermediate   it

we can drop the three, ids [8268,8018,8021] and can have the rest.

In [35]:
data.drop( [8268,8018,8021],axis=0,inplace=True)

In [36]:
data.Edition_Language.replace('None','English',inplace=True)

In [37]:
data.Edition_Language.value_counts()

English    10505
Name: Edition_Language, dtype: int64

We finally have 10505 English Books.

In [38]:
none.Book_Title.tolist()

['Reached',
 'Opal',
 'The Dreamers',
 'The Gift',
 'Queen Song',
 'Head On',
 'The Walking Dead, Vol. 11: Fear the Hunters',
 'Life 3.0: Being Human in the Age of Artificial Intelligence',
 'Sea of Silver Light',
 'Beyond Shame',
 'After the Flood',
 'The King in Yellow',
 'We Set the Dark on Fire',
 'Mavericks',
 'Rebellion',
 'Dog Man: Lord of the Fleas',
 'The King of Crows',
 'Beggars in Spain',
 'Traitor Born',
 'Executed',
 'Points of Impact',
 'The Walled City',
 'Earthcore',
 'Change Agent',
 'Marking Time',
 'The Legion of Flame',
 'The New Wilderness',
 'Generation One',
 'Counterstrike',
 'Ghost Country',
 'The Wolves of Winter',
 'The Summer of Impossible Things',
 'Vivian Apple at the End of the World',
 'After the Ending',
 'Dog Man: For Whom the Ball Rolls',
 'Valkyrie',
 'Before the Broken Star',
 'Legionnaire',
 'Clockwork Boys',
 'A Big Ship at the Edge of the Universe',
 'Hazards of Time Travel',
 'Capturing Cara',
 'Captive',
 'Extinct',
 'Archangel One',
 'Playing

## Descriptions

How long are the descriptions?\
Here is the word lenght of the Book_Descriptiom

In [39]:
data['length']=[len(d.split()) for d in data.Book_Description.tolist()]

In [40]:
data.head()

Book_Title  \
ID                                            
2606  Harry Potter and the Sorcerer's Stone   
8151                       The Hunger Games   
7832                                   1984   
2009                              Divergent   
2534    The Hobbit, or There and Back Again   

                           Original_Book_Title      Author_Name  \
ID                                                                
2606  Harry Potter and the Philosopher's Stone     J.K. Rowling   
8151                          The Hunger Games  Suzanne Collins   
7832                      Nineteen Eighty-Four    George Orwell   
2009                                 Divergent    Veronica Roth   
2534       The Hobbit, or There and Back Again   J.R.R. Tolkien   

     Edition_Language  Rating_score  Rating_votes  Review_number  \
ID                                                                 
2606          English          4.48       7336299         116268   
8151          English          4.33       6572958         174090   
7832          English          4.19       3277088          73134   
2009          English          4.19       3008156         109103   
2534          English          4.28       2990501          49914   

                                       Book_Description  Year_published  \
ID                                                                        
2606  Harry Potter's life is miserable. His parents ...            1997   
8151  Could you survive on your own in the wild, wit...            2008   
7832  Among the seminal texts of the 20th century, N...            1949   
2009  In Beatrice Prior's dystopian Chicago world, s...            2011   
2534  In a hole in the ground there lived a hobbit. ...            1937   

                                                 Genres  \
ID                                                        
2606  {'Fantasy': 63272, 'Fiction': 18715, 'Young Ad...   
8151  {'Young Adult': 31089, 'Fiction': 17571, 'Scie...   
7832  {'Classics': 36120, 'Fiction': 22533, 'Science...   
2009  {'Young Adult': 20861, 'Science Fiction (Dysto...   
2534  {'Fantasy': 52370, 'Classics': 17642, 'Fiction...   

                                                    url  length  
ID                                                               
2606  https://www.goodreads.com/book/show/3.Harry_Po...     220  
8151  https://www.goodreads.com/book/show/2767052-th...     164  
7832  https://www.goodreads.com/book/show/40961427-1984      96  
2009  https://www.goodreads.com/book/show/13335037-d...     216  
2534  https://www.goodreads.com/book/show/5907.The_H...     147

In [41]:
data[data.length.isin(range(0,10))][["Book_Title",'Book_Description','length']]\
.sort_values(by=['length'],ascending=True).head()

Book_Title  \
ID                                                        
8455                                    Into the Future   
925                                      Eight's Origin   
652              Obsidian: Last Chapter in Daemon's POV   
8003            2001 Nights: The Death Trilogy Overture   
1845  The Collected What If? Eminent Historians Imag...   

                              Book_Description  length  
ID                                                      
8455                              intermediate       1  
925                              \n  [image]\n       1  
652   http://www.scribd.com/doc/108904697/I...       1  
8003                          SCIENCE FICTION.       2  
1845                                  What If?       2

Removing the books having description as less than 10 words.

In [42]:
data.drop(data.index[data.length.isin(range(0,10))],axis=0,inplace=True)

In [43]:
data.shape[0]

10481

## Genres

considering 4-5 top genres from the dictionary and we have two options:
1. enter the genre string into the description text.
2. take percentage of genre and take it as a numerical feature.

But, proceeding with the first option and adding top 5 genres into the description

In [44]:
def genre_text_extraction(text):
    genre_list=re.sub(r"[\d:{}'']","",text).split(',')
    genre_list=genre_list[:5]
    return ' '.join(genre_list)

In [45]:
data.Genres=data.Genres.map(genre_text_extraction)

In [46]:
data[['Genres']].head()

Genres
ID                                                     
2606  Fantasy   Fiction   Young Adult   Fantasy (Mag...
8151  Young Adult   Fiction   Science Fiction (Dysto...
7832  Classics   Fiction   Science Fiction   Science...
2009  Young Adult   Science Fiction (Dystopia)    Fi...
2534  Fantasy   Classics   Fiction   Adventure   You...

In [47]:
data.head()

Book_Title  \
ID                                            
2606  Harry Potter and the Sorcerer's Stone   
8151                       The Hunger Games   
7832                                   1984   
2009                              Divergent   
2534    The Hobbit, or There and Back Again   

                           Original_Book_Title      Author_Name  \
ID                                                                
2606  Harry Potter and the Philosopher's Stone     J.K. Rowling   
8151                          The Hunger Games  Suzanne Collins   
7832                      Nineteen Eighty-Four    George Orwell   
2009                                 Divergent    Veronica Roth   
2534       The Hobbit, or There and Back Again   J.R.R. Tolkien   

     Edition_Language  Rating_score  Rating_votes  Review_number  \
ID                                                                 
2606          English          4.48       7336299         116268   
8151          English          4.33       6572958         174090   
7832          English          4.19       3277088          73134   
2009          English          4.19       3008156         109103   
2534          English          4.28       2990501          49914   

                                       Book_Description  Year_published  \
ID                                                                        
2606  Harry Potter's life is miserable. His parents ...            1997   
8151  Could you survive on your own in the wild, wit...            2008   
7832  Among the seminal texts of the 20th century, N...            1949   
2009  In Beatrice Prior's dystopian Chicago world, s...            2011   
2534  In a hole in the ground there lived a hobbit. ...            1937   

                                                 Genres  \
ID                                                        
2606  Fantasy   Fiction   Young Adult   Fantasy (Mag...   
8151  Young Adult   Fiction   Science Fiction (Dysto...   
7832  Classics   Fiction   Science Fiction   Science...   
2009  Young Adult   Science Fiction (Dystopia)    Fi...   
2534  Fantasy   Classics   Fiction   Adventure   You...   

                                                    url  length  
ID                                                               
2606  https://www.goodreads.com/book/show/3.Harry_Po...     220  
8151  https://www.goodreads.com/book/show/2767052-th...     164  
7832  https://www.goodreads.com/book/show/40961427-1984      96  
2009  https://www.goodreads.com/book/show/13335037-d...     216  
2534  https://www.goodreads.com/book/show/5907.The_H...     147

Let us conactenate the following text columns into one:
- Original Book Title
- Author Name
- Book Description
- Genres

In [48]:
data['text']=data.Original_Book_Title+" "+data.Author_Name+" "+\
                data.Book_Description+" "+data.Genres

In [49]:
df=data.drop(['Original_Book_Title','Edition_Language','Rating_votes',\
              'Review_number','Genres','length'],axis=1)

In [50]:
df.head()

Book_Title      Author_Name  Rating_score  \
ID                                                                           
2606  Harry Potter and the Sorcerer's Stone     J.K. Rowling          4.48   
8151                       The Hunger Games  Suzanne Collins          4.33   
7832                                   1984    George Orwell          4.19   
2009                              Divergent    Veronica Roth          4.19   
2534    The Hobbit, or There and Back Again   J.R.R. Tolkien          4.28   

                                       Book_Description  Year_published  \
ID                                                                        
2606  Harry Potter's life is miserable. His parents ...            1997   
8151  Could you survive on your own in the wild, wit...            2008   
7832  Among the seminal texts of the 20th century, N...            1949   
2009  In Beatrice Prior's dystopian Chicago world, s...            2011   
2534  In a hole in the ground there lived a hobbit. ...            1937   

                                                    url  \
ID                                                        
2606  https://www.goodreads.com/book/show/3.Harry_Po...   
8151  https://www.goodreads.com/book/show/2767052-th...   
7832  https://www.goodreads.com/book/show/40961427-1984   
2009  https://www.goodreads.com/book/show/13335037-d...   
2534  https://www.goodreads.com/book/show/5907.The_H...   

                                                   text  
ID                                                       
2606  Harry Potter and the Philosopher's Stone J.K. ...  
8151  The Hunger Games Suzanne Collins Could you sur...  
7832  Nineteen Eighty-Four George Orwell Among the s...  
2009  Divergent Veronica Roth In Beatrice Prior's dy...  
2534  The Hobbit, or There and Back Again J.R.R. Tol...

Having proper indices for books.

In [51]:
df['Id']=[i for i in range(0,df.shape[0])]

In [52]:
#df.set_index('Id',inplace=True)

In [53]:
df.head()

Book_Title      Author_Name  Rating_score  \
ID                                                                           
2606  Harry Potter and the Sorcerer's Stone     J.K. Rowling          4.48   
8151                       The Hunger Games  Suzanne Collins          4.33   
7832                                   1984    George Orwell          4.19   
2009                              Divergent    Veronica Roth          4.19   
2534    The Hobbit, or There and Back Again   J.R.R. Tolkien          4.28   

                                       Book_Description  Year_published  \
ID                                                                        
2606  Harry Potter's life is miserable. His parents ...            1997   
8151  Could you survive on your own in the wild, wit...            2008   
7832  Among the seminal texts of the 20th century, N...            1949   
2009  In Beatrice Prior's dystopian Chicago world, s...            2011   
2534  In a hole in the ground there lived a hobbit. ...            1937   

                                                    url  \
ID                                                        
2606  https://www.goodreads.com/book/show/3.Harry_Po...   
8151  https://www.goodreads.com/book/show/2767052-th...   
7832  https://www.goodreads.com/book/show/40961427-1984   
2009  https://www.goodreads.com/book/show/13335037-d...   
2534  https://www.goodreads.com/book/show/5907.The_H...   

                                                   text  Id  
ID                                                           
2606  Harry Potter and the Philosopher's Stone J.K. ...   0  
8151  The Hunger Games Suzanne Collins Could you sur...   1  
7832  Nineteen Eighty-Four George Orwell Among the s...   2  
2009  Divergent Veronica Roth In Beatrice Prior's dy...   3  
2534  The Hobbit, or There and Back Again J.R.R. Tol...   4

Stop words

In [54]:
stop = get_stop_words('en')

# remove non-alphanumeric, non-space
stop = [re.sub(r'([^\s\w]|_)+', '', x) for x in stop]

In [55]:
def remove_stop(text, stop):
    """
    As function suggests,
    checks fro stop word in the text and doesnt append stop word
    Returns the text without stop words
    """
    custom_stop = stop
    
    #appending a word if it is not a stop word
    new_text = []
    for word in text:
        if word not in custom_stop:
            new_text.append(word)
    return new_text

In [56]:
tokenizer = RegexpTokenizer(r'\w+')
p_stemmer = PorterStemmer()
l_stemmer = WordNetLemmatizer()

In [57]:
def stem_list(text, p_stemmer):
    """
    Does stemming and lemmataization of the words in the text
    Returns a list of those words
    """
    new_list = []
    for word in text:
        new_list.append(p_stemmer.stem(word))
    return new_list

def lem_list(text, l_stemmer):
    new_list = []
    for word in text:
        new_list.append(l_stemmer.lemmatize(word))
    return new_list

In [58]:
def preprocess_text(text):
    
    """
    FLOW-
    1. Remove the unwanted characters
    2. Remove the non alphanueric, non space characters
    4. Remove stop words from the text
    5. Lemmatize the words
    6. Join the words in the list and retrun text
    """
    # remove mixed alphanumeric
    text = re.sub(r"""(?x) # verbose regex
                            \b    # Start of word
                            (?=   # Look ahead to ensure that this word contains...
                             \w*  # (after any number of alphanumeric characters)
                             \d   # ...at least one digit.
                            )     # End of lookahead
                            \w+   # Match the alphanumeric word
                            \s*   # Match any following whitespace""", 
                             "", text)
    
    # remove non-alphanumeric, non-space
    text = re.sub(r'([^\s\w]|_)+', '', text)
    
    # tokenize text
    text = tokenizer.tokenize(text.lower())
    
    # remove stop words
    text = remove_stop(text, stop)
    
    # stem
    text = lem_list(text, l_stemmer)

    
    new_text = ' '.join(text)
    
    
    return new_text

In [59]:
df['text']=df.text.map(preprocess_text)

In [60]:
df['text'].head(2).tolist()

['harry potter philosopher stone jk rowling harry potter life miserable parent dead stuck heartless relative force live tiny closet stair fortune change receives letter tell truth wizard mysterious visitor rescue relative take new home hogwarts school witchcraft wizardryafter lifetime bottling magical power harry finally feel like normal kid even within wizarding community special boy lived person ever survived killing curse inflicted evil lord voldemort launched brutal takeover wizarding world vanish failing kill harrythough harrys first year hogwarts best life everything perfect dangerous secret object hidden within castle wall harry belief responsibility prevent falling evil hand will bring contact force terrifying ever imaginedfull sympathetic character wildly imaginative situation countless exciting detail first installment series assembles unforgettable magical world set stage many highstakes adventure come fantasy fiction young adult fantasy magic childrens',
 'hunger game suzan

In [61]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [62]:
cv=CountVectorizer()
cv_matrix=cv.fit_transform(df.text)
cv_cosine_sim=cosine_similarity(cv_matrix)

In [63]:
def get_title_from_index(id):
    """
    Returns the title of the bOOK given the index
    """
    return df[df["Id"]==id]["Book_Title"].values[0]

def get_index_from_title(title):
    """
    Returns the index of the Book given the title
    """
    return df[df.Book_Title == title].Id.values[0]

def get_recommendations(book_id, sim_matrix):
    """
    -Store the recommendations in a list
    -Get the title of book using get_title_from_index
    -Genarate cosine similarity matrix for the book and enumerate it
    -Sort the list based on the similarity
    -Get the index of the book (due to enumeration) 
    and get corresponding title
    -Retrun the list of recs
    """
    recommendations = list()
    
    podcast_title = get_title_from_index(book_id)
    similar_books =  list(enumerate(sim_matrix[book_id]))
    sorted_similar_books = sorted(similar_books,key=lambda x:x[1],reverse=True)
    
    for i in range(4):
        title = get_title_from_index(sorted_similar_books[i][0])
        recommendations.append(title)
    
    return recommendations

def get_dissimilar_recommendations(book_id, sim_matrix):
    """
    -Store the recommendations in a list
    -Get the title of book using get_title_from_index
    -Genarate cosine similarity matrix for the book and enumerate it
    -Sort the list based on the similarity
    -Get the index of the book (due to enumeration) 
    and get corresponding title
    -Retrun the list of recs
    """
    recommendations = list()
    
    podcast_title = get_title_from_index(book_id)
    similar_books =  list(enumerate(sim_matrix[book_id]))
    sorted_similar_books = sorted(similar_books,key=lambda x:x[1])
    
    for i in range(4):
        title = get_title_from_index(sorted_similar_books[i][0])
        recommendations.append(title)
    
    return recommendations

In [64]:
book="Harry Potter and the Sorcerer's Stone"
get_recommendations(get_index_from_title(book),cv_cosine_sim)

["Harry Potter and the Sorcerer's Stone",
 'Harry Potter and the Chamber of Secrets',
 'Harry Potter and the Half-Blood Prince',
 'Harry Potter and the Goblet of Fire']

In [65]:
book="The Hunger Games"
get_recommendations(get_index_from_title(book),cv_cosine_sim)

['The Hunger Games',
 'Catching Fire',
 'The Uploaded',
 'The Hunger Games Trilogy Boxset']

In [66]:
book="1984"
get_recommendations(get_index_from_title(book),cv_cosine_sim)

['1984', 'We', 'Dr. Strangelove', 'Redliners']

In [67]:
book="Harry Potter and the Sorcerer's Stone"
get_dissimilar_recommendations(get_index_from_title(book),cv_cosine_sim)

['The Hacker Crackdown: Law and Disorder on the Electronic Frontier',
 'Invincible: Ultimate Collection, Vol. 7',
 'Invincible: Ultimate Collection, Vol. 8',
 'Hanna-Barbera Cartoons']

In [68]:
book="1984"
get_dissimilar_recommendations(get_index_from_title(book),cv_cosine_sim)

['City of Fallen Angels',
 'The Unbecoming of Mara Dyer',
 'The Iron Queen',
 'Magic Burns']

TFIDF

In [69]:
from sklearn.feature_extraction.text import TfidfVectorizer

tf=TfidfVectorizer()
tf_matrix=tf.fit_transform(df["text"])
tf_cosine_sim=cosine_similarity(tf_matrix)

In [70]:
book="1984"
get_recommendations(get_index_from_title(book),tf_cosine_sim)

['1984', 'We', 'Animal Farm / 1984', 'Animal Farm']

In [71]:
book="Harry Potter and the Sorcerer's Stone"
get_recommendations(get_index_from_title(book),tf_cosine_sim)

["Harry Potter and the Sorcerer's Stone",
 'Harry Potter and the Chamber of Secrets',
 'Harry Potter and the Goblet of Fire',
 'Harry Potter and the Half-Blood Prince']